pkg/controllers/subnet/subnet_controller.go

```diff
@@ -50,102 +48,141 @@ type SubnetReconciler struct {
-			if err := r.Client.Update(ctx, obj); err != nil {
-				log.Error(err, "deletion failed, would retry exponentially", "subnet", req.NamespacedName)
-				deleteFail(r, ctx, obj, err.Error())
-				return ResultRequeue, err
-			}
-			log.V(1).Info("removed finalizer", "subnet", req.NamespacedName)
-			deleteSuccess(r, ctx, obj)
-		} else {
-			log.Info("finalizers cannot be recognized", "subnet", req.NamespacedName)
+		log.Info("Updated Subnet CR", "Subnet", req.NamespacedName)
+	}
+
+	tags := r.SubnetService.GenerateSubnetNSTags(subnetCR)
+	if tags == nil {
+		log.Error(nil, "failed to generate Subnet tags", "Subnet", req.NamespacedName)
+		return ResultRequeue, errors.New("failed to generate Subnet tags")
+	}
+	// List VPC Info
+	vpcInfoList := r.VPCService.ListVPCInfo(req.Namespace)
+	if len(vpcInfoList) == 0 {
+		log.Info("No VPC info found, requeueing", "Namespace", req.Namespace)
+		return ResultRequeueAfter10sec, nil
+	}
+	// Create or update the subnet in NSX
+	if _, err := r.SubnetService.CreateOrUpdateSubnet(subnetCR, vpcInfoList[0], tags); err != nil {
+		if errors.As(err, &nsxutil.ExceedTagsError{}) {
+			log.Error(err, "tags limit exceeded, not retrying", "Subnet", req.NamespacedName)
+			updateFail(r, ctx, subnetCR, err.Error())
+			return ResultNormal, nil
 		}
+		log.Error(err, "failed to create/update Subnet, retrying", "Subnet", req.NamespacedName)
+		updateFail(r, ctx, subnetCR, err.Error())
+		return ResultRequeue, err
+	}
+	// Update status
+	if err := r.updateSubnetStatus(subnetCR); err != nil {
+		log.Error(err, "failed to update Subnet status, retrying", "Subnet", req.NamespacedName)
+		updateFail(r, ctx, subnetCR, err.Error())
+		return ResultRequeue, err
 	}
+	updateSuccess(r, ctx, subnetCR)
 	return ctrl.Result{}, nil
 }
 
-func (r *SubnetReconciler) DeleteSubnet(obj v1alpha1.Subnet) error {
-	nsxSubnets := r.SubnetService.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetCRUID, string(obj.GetUID()))
-	if len(nsxSubnets) == 0 {
-		log.Info("no subnet found for subnet CR", "uid", string(obj.GetUID()))
-		return nil
-	}
-	portNums := len(r.SubnetPortService.GetPortsOfSubnet(*nsxSubnets[0].Id))
-	if portNums > 0 {
-		err := errors.New("subnet still attached by port")
-		log.Error(err, "", "ID", *nsxSubnets[0].Id)
+func (r *SubnetReconciler) deleteSubnetByID(subnetID string) error {
+	nsxSubnets := r.SubnetService.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetCRUID, subnetID)
+	return r.deleteSubnets(nsxSubnets)
+}
+
+func (r *SubnetReconciler) deleteSubnets(nsxSubnets []*model.VpcSubnet) error {
+	crdSubnetIDs, err := r.listSubnetCRs(context.Background())
+	if err != nil {
+		log.Error(err, "failed to list Subnet CRs")
 		return err
 	}
-	return r.SubnetService.DeleteSubnet(*nsxSubnets[0])
+	crdSubnetIDsSet := sets.NewString(crdSubnetIDs...)
+	for _, nsxSubnet := range nsxSubnets {
+		uid := nsxutil.FindTag(nsxSubnet.Tags, servicecommon.TagScopeSubnetCRUID)
```

Comment:
Why we need to compare UID again here?

Reply:
This comparing is needed to check if K8s CR of UID is still remaining  in K8s or not. So, we can know which CR is created or deleted in case that the CRs has the same ns/name.